quickjs.c

```diff
@@ -9261,6 +9304,56 @@ static int js_shape_prepare_update(JSContext *ctx, JSObject *p,
     return 0;
 }
 
+int js_shape_delete_watchpoints(JSRuntime *rt, JSShape *shape, JSInlineCacheRingSlot *target) {
+    struct list_head *el, *el1;
+    if (unlikely(!shape || !shape->watchpoint))
+        goto end;
+    list_for_each_safe(el, el1, shape->watchpoint) {
+        JSInlineCacheWatchpoint *o = list_entry(el, JSInlineCacheWatchpoint, link);
+            if (!ic_watchpoint_delete_handler(rt, o->ref, o->offset, o->atom, target)) {
+                list_del(el);
+                js_free_rt(rt, o);
+                break;
+            }
+    }
+end:
+    return 0;
+}
+
+int js_shape_free_watchpoints(JSRuntime *rt, JSShape *shape) {
+    struct list_head *el, *el1;
+    if (unlikely(!shape || !shape->watchpoint))
+        goto end;
+    list_for_each_safe(el, el1, shape->watchpoint) {
+        JSInlineCacheWatchpoint *o = list_entry(el, JSInlineCacheWatchpoint, link);
+        ic_watchpoint_free_handler(rt, o->ref, o->offset, o->atom);
+        list_del(el);
+        js_free_rt(rt, o);
+    }
+    list_empty(shape->watchpoint);
+    js_free_rt(rt, shape->watchpoint);
+end:
+    return 0;
+}
+
+JSInlineCacheWatchpoint *js_shape_create_watchpoint(JSRuntime *rt, JSShape *shape, JSInlineCacheRingSlot *ptr, uint8_t offset, JSAtom atom)
+{
+    JSInlineCacheWatchpoint *o;
+    o = js_malloc_rt(rt, sizeof(*o));
+    if (unlikely(!o))
+        return NULL;
+    o->ref = ptr;
+    o->atom = atom;
+    o->offset = offset;
+    if (!shape->watchpoint) {
+        shape->watchpoint = js_malloc_rt(rt, sizeof(*shape->watchpoint));
```

Comment:
You should handle allocation failure here, but I would suggest turning `watchpoint` from a `struct list_head *` to an embedded `struct list_head`. Initialize it with `init_list_head(&sh->watchpoint)` in js_new_shape2.

That way, you also don't have to check for nullity before iterating in js_shape_free_watchpoints and js_shape_delete_watchpoints.